future-apps/agent/contracts/Agent.sol

```diff
@@ -0,0 +1,207 @@
+    {
+        // Prevent an infinite loop by setting the app itself as its designated signer.
+        // An undetectable loop can be created by setting a different contract as the
+        // designated signer which calls back into `isValidSignature`.
+        // Given that `isValidSignature` is always called with just 50k gas, the max
+        // damage of the loop is wasting 50k gas.
+        require(_designatedSigner != address(this), ERROR_DESIGNATED_TO_SELF);
+
+        address oldDesignatedSigner = designatedSigner;
+        designatedSigner = _designatedSigner;
+
+        emit SetDesignatedSigner(msg.sender, oldDesignatedSigner, _designatedSigner);
+    }
+
+    /**
+    * @notice Pre-sign hash `_hash`
+    * @param _hash Hash that will be considered signed regardless of the signature checked with 'isValidSignature()'
+    */
+    function presignHash(bytes32 _hash)
+        external
+        authP(ADD_PRESIGNED_HASH_ROLE, arr(_hash))
+    {
+        isPresigned[_hash] = true;
+
+        emit PresignHash(msg.sender, _hash);
+    }
+
+    function isForwarder() external pure returns (bool) {
+        return true;
+    }
+
+    function supportsInterface(bytes4 interfaceId) external pure returns (bool) {
+        return
+            interfaceId == ISVALIDSIG_INTERFACE_ID ||
+            interfaceId == EIP165_SUPPORT_INTERFACE_ID;
+    }
+
+    /**
+    * @notice Execute the script as the Agent app
+    * @dev IForwarder interface conformance. Forwards any token holder action.
+    * @param _evmScript Script being executed
+    */
+    function forward(bytes _evmScript)
+        public
+        authP(RUN_SCRIPT_ROLE, arr(getScriptACLParam(_evmScript)))
+    {
+        bytes memory input = ""; // no input
+        address[] memory blacklist = new address[](0); // no addr blacklist, can interact with anything
+        runScript(_evmScript, input, blacklist);
+        // We don't need to emit an event here as EVMScriptRunner will emit ScriptResult if successful
+    }
+
+    function isValidSignature(bytes data, bytes signature) public view returns (bool) {
+        return isValidSignature(keccak256(data), signature);
+    }
+
+    function isValidSignature(bytes32 hash, bytes signature) public view returns (bool) {
+        // Short-circuit in case the hash was presigned. Optimization as performing calls
+        // and ecrecover is more expensive than an SLOAD.
+        if (isPresigned[hash]) {
+            return true;
+        }
+
+        // Checks if designatedSigner is a contract, and if it supports the isValidSignature interface
+        if (safeSupportsInterface(IERC165(designatedSigner), ISVALIDSIG_INTERFACE_ID)) {
+            // designatedSigner.isValidSignature(hash, signature) as a staticall
+            IERC1271 signerContract = IERC1271(designatedSigner);
+            bytes memory calldata = abi.encodeWithSelector(signerContract.isValidSignature.selector, hash, signature);
```

Comment:
I would rename this to `calldata_` or something like this. I was having problems with truffle 5 importing this as a dependency. Same for the 2 other occurrences below.

Reply:
Actually surprised this compiles; `calldata` should be a reserved word, no?

Reply:
Yes, it's surprising, I guess it's using the context or something. Anyway, it doesn't work with Truffle 5, and it could be confusing.